airbyte-cdk/python/airbyte_cdk/entrypoint.py

```diff
@@ -160,8 +162,38 @@ def read(
         if self.source.check_config_against_spec:
             self.validate_connection(source_spec, config)
 
-        yield from self.source.read(self.logger, config, catalog, state)
-        yield from self._emit_queued_messages(self.source)
+        stream_message_count: DefaultDict[HashableStreamDescriptor, int] = defaultdict(int)
+        for message in self.source.read(self.logger, config, catalog, state):
+            yield self.handle_record_counts(message, stream_message_count)
+        for message in self._emit_queued_messages(self.source):
+            yield self.handle_record_counts(message, stream_message_count)
+
+    @staticmethod
+    def handle_record_counts(message: AirbyteMessage, stream_message_count: DefaultDict[HashableStreamDescriptor, int]) -> AirbyteMessage:
+        if message.type == Type.RECORD:
+            logger.debug(f"Message type: {message}")
+            descriptor = HashableStreamDescriptor(name=message.record.stream, namespace=message.record.namespace)
+            stream_message_count[descriptor] += 1
+        elif message.type == Type.STATE:
+            logger.debug(f"Message type: {message}")
+            if not message.state.stream:
+                raise ValueError("State message was not in per-stream state format which is required for record counts")
+
+            descriptor = HashableStreamDescriptor(
+                name=message.state.stream.stream_descriptor.name, namespace=message.state.stream.stream_descriptor.namespace
+            )
+            source_stats_for_stream = message.state.sourceStats or AirbyteStateStats()
+            source_stats_for_stream.recordCount = stream_message_count.get(descriptor, 0)
+            message.state.sourceStats = source_stats_for_stream
+
+            # todo: remove after pre-release testing
```

Comment:
Temporary just to make it easy to validate our state counts in pre-release versions